sdk/dotnet/Pulumi.Automation/Runtime/LanguageRuntimeService.cs

```diff
@@ -23,10 +24,69 @@ public LanguageRuntimeService(CallerContext callerContext)
+        public override async Task<GetRequiredPluginsResponse> GetRequiredPlugins(GetRequiredPluginsRequest request, ServerCallContext context)
         {
             var response = new GetRequiredPluginsResponse();
-            return Task.FromResult(response);
+            var assembly = this._callerContext.Program.DiscoveryAssembly;
+
+            // if discovery assembly is null, assuming consumer installed plugins manually
+            if (assembly is null)
+                return response;
+
+            var pulumiAssemblyNames = assembly.GetReferencedAssemblies().Where(x => x.Name != null && x.Name.StartsWith("Pulumi.")).ToArray();
+            foreach (var assemblyName in pulumiAssemblyNames)
+            {
+                var pulumiAssembly = Assembly.Load(assemblyName);
+
+                // get version.txt
+                // this part would probably be less error prone if there was an assembly attribute available
+                // instead of using the embedded version.txt resource
+                // pulumiAssembly.GetCustomAttribute<ResourcePluginAttribute>();
+                var resources = pulumiAssembly.GetManifestResourceNames();
+                var versionResource = resources.FirstOrDefault(x => x.EndsWith("version.txt"));
+                if (versionResource is null)
+                {
+                    // no version.txt so is not resource plugin
+                    continue;
+                }
+
+                using var versionStream = pulumiAssembly.GetManifestResourceStream(versionResource);
+                if (versionStream is null)
+                {
+                    // we've already verified that version.txt should be present so throw here?
+                    throw new ApplicationException($"Unable to load version.txt from {assemblyName.FullName}");
+                }
+
+                var versionReader = new StreamReader(versionStream);
+                var versionText = await versionReader.ReadToEndAsync().ConfigureAwait(false);
+
+                // ToLower() everything after "Pulumi." to determine plugin name
+                var pluginName = assemblyName.Name!.Substring("Pulumi.".Length).ToLower();
+                var version = versionText.Trim();
+                var parts = versionText.Split('\n', StringSplitOptions.RemoveEmptyEntries)
+                    .Select(x => x.Trim())
+                    .ToArray();
+
+                // if version.txt is 2 lines than it is specifying plugin name explicitly
+                if (parts.Length is 2)
+                {
+                    pluginName = parts[0];
+                    version = parts[1];
+                }
+
+                // pre-pend 'v' if it isn't present on version string
+                if (!version.StartsWith('v'))
+                    version = $"v{version}";
+
+                response.Plugins.Add(new PluginDependency
```

Comment:
And finally, should we consider any caching of `DiscoveryAssembly -> RequiredPlugins` to cut down on how often we might be loading the same assemblies, or would that be pre-optimization?